### ISSUE TITLE
Factorial & Binomial Coefficient

### ISSUE THREAD

**Issue by juliusbierk** [pytorch/pytorch#47841]
## 🚀 Feature
`torch.factorial` which behaves similar to `scipy.special.factorial`
`torch.binom` which behaves similar to `scipy.special.binom`

## Motivation & Pitch
torch currenltly only offers `lgamma` which can be used to calculate factorial as `torch.exp(torch.lgamma(x + 1))`, this is a bit cumbersome for simple mathematical operations, which are often used.

The binomial coefficient can currently be implemented as `torch.exp(torch.lgamma(n + 1) - torch.lgamma((n - k) + 1) - torch.lgamma(k + 1))`, but this has the problem when `n < k` in `binom(n, k)`. In case the result will correctly be returned as  `torch.exp(-inf) = 0`, but gradients will be `nan`.

## Alternatives

The binomial coefficient can currently be implemented by something like
```
def torch_binom(n, k):
    mask = n.detach() >= k.detach()
    n = mask * n
    k = mask * k
    a = torch.lgamma(n + 1) - torch.lgamma((n - k) + 1) - torch.lgamma(k + 1)
    return torch.exp(a) * mask
```

which will allow gradients to the propagated through. But this is very cumbersome for such a simple function.

## Additional context

I you can guide me in the right direction, I would not mind adding this if you agree that it would be a useful feature.

**Comment by mruberry** [comment]
Hey @juliusbierk, thank you for the recommendation. We would accept two PRs, one for each of these functions. The signature of each should be consistent with SciPy plus support the out= argument.

This PR adding gcd and lcm may be a good example of how to proceed: https://github.com/pytorch/pytorch/pull/40651. Let me know if you have additional questions.
